src/vector.rs

```diff
@@ -283,6 +298,32 @@ where
     {
         self.square_length().sqrt()
     }
+
+    /// Returns this vector projected onto another one.
+    #[inline]
+    pub fn project_onto(&self, onto: Self) -> Self
+    where
+        T: Div<T, Output = T>
+    {
+        onto * (self.dot(onto) / onto.square_length())
+    }
+}
+
+impl<T, U> Vector2D<T, U>
+where
+    T: Copy + Mul<T, Output = T> + Add<T, Output = T> + Sub<T, Output = T>
+    + PartialOrd + Float
+{
+    /// Return this vector capped to a maximum length.
+    #[inline]
+    pub fn cap_length(&self, max_length: T) -> Self {
+        let square_length = self.square_length();
+        if square_length > max_length * max_length {
+            return self.normalize() * max_length;
```

Comment:
should we re-use the `square_length` here?

Reply:
I was hoping that with everything is marked #[inline] the compiler would have an easy time factoring it but I admit I didn't bother to check.
I'll change make it reuse `square_length` so that we don't have to think about it.